internal/speakerlist/speakerlist.go

```diff
@@ -0,0 +1,301 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package speakerlist
+
+import (
+	"crypto/sha256"
+	golog "log"
+	"strconv"
+	"sync"
+	"time"
+
+	"go.universe.tf/metallb/internal/k8s"
+
+	gokitlog "github.com/go-kit/kit/log"
+	"github.com/hashicorp/memberlist"
+)
+
+// SpeakerList gives you the list of healthy speaker
+type SpeakerList struct {
+	l         gokitlog.Logger
+	client    *k8s.Client
+	stopCh    chan struct{}
+	namespace string
+	labels    string
+
+	// following fields are nil when MemberList is disabled
+	mlEventCh chan memberlist.NodeEvent
+	mList     *memberlist.Memberlist
+	mlJoinCh  chan struct{}
+
+	mlMux         sync.Mutex // Mutex for mlSpeakersIPs
+	mlSpeakersIPs []string   // Speaker IPs
+}
+
+// New creates a new SpeakerList
+func New(logger gokitlog.Logger, nodeName, bindAddr, bindPort, secret, namespace, labels string, stopCh chan struct{}) (*SpeakerList, error) {
+	sl := SpeakerList{
+		l:         logger,
+		stopCh:    stopCh,
```

Comment:
I agree. Not sure if it should be separate or not, though, as code wasn't very readable on my quick try, but maybe I was trying too quick. I left that for a following PR, though.

When I opened this PR [I created an issue][link] about leaving orphan goroutines. Currently the code on master is not doing anything about it. I think we should also create a wait group and actually wait for them.

[link]: https://github.com/metallb/metallb/issues/663

Reply:
I added the issue to the 0.9.4 milestone.

Reply:
@johananl as I mentioned in the comments, this PR **is** adding new go routines that are leaked. Cleaning them up is the same effort as fixing other leaking go routines in the speakerlist code.

I don't like adding new leaks either, but I created the issue for that reason, as I think this will be better fixed when this code is improved. Most go routines might not exist then, as we might use a simpler watch mechanism.

IMHO, at this point in time, I'd really like to release a patch release soon and work on further improvements for the next minor. IOW, I prefer to fix the leaking go routines for the 0.10 release.

WDYT?

Reply:
@johananl ping?

Reply:
I'd like to get this released ASAP, too, but I also feel uncomfortable about consciously introducing leaks. Should I take a quick look at this and try to solve the problem?

Reply:
The following seems to work:

```diff
diff --git a/internal/speakerlist/speakerlist.go b/internal/speakerlist/speakerlist.go
index bd8c95c1..7cd31443 100644
--- a/internal/speakerlist/speakerlist.go
+++ b/internal/speakerlist/speakerlist.go
@@ -27,9 +27,16 @@ import (
 
 // SpeakerList represents a list of healthy speakers.
 type SpeakerList struct {
-	l         gokitlog.Logger
-	client    *k8s.Client
-	stopCh    chan struct{}
+	l      gokitlog.Logger
+	client *k8s.Client
+
+	// A channel callers can use to stop the SpeakerList.
+	stopCh chan struct{}
+	// A channel SpeakerList uses to stop child goroutines.
+	childrenStopCh chan struct{}
+	// A WaitGroup for tracking child goroutines.
+	childrenWG sync.WaitGroup
+
 	namespace string
 	labels    string
 
@@ -45,10 +52,11 @@ type SpeakerList struct {
 // New creates a new SpeakerList and returns a pointer to it.
 func New(logger gokitlog.Logger, nodeName, bindAddr, bindPort, secret, namespace, labels string, stopCh chan struct{}) (*SpeakerList, error) {
 	sl := SpeakerList{
-		l:         logger,
-		stopCh:    stopCh,
-		namespace: namespace,
-		labels:    labels,
+		l:              logger,
+		stopCh:         stopCh,
+		childrenStopCh: make(chan struct{}),
+		namespace:      namespace,
+		labels:         labels,
 	}
 
 	if namespace == "" || labels == "" || bindAddr == "" {
@@ -143,24 +151,27 @@ func (sl *SpeakerList) Start(client *k8s.Client) {
 	sl.mlSpeakerIPs = iplist
 	sl.mlMux.Unlock()
 
-	// Update mlSpeakerIPs in the background.
-	go sl.updateSpeakerIPs()
+	sl.childrenWG.Add(3)
 
-	go sl.memberlistWatchEvents()
-	go sl.joinMembers()
+	// Update mlSpeakerIPs in the background.
+	go sl.updateSpeakerIPs(sl.childrenStopCh)
+	go sl.memberlistWatchEvents(sl.childrenStopCh)
+	go sl.joinMembers(sl.childrenStopCh)
 }
 
 // updateSpeakerIPs runs forever updating the sl.mlSpeakerIPs slice with the
 // IPs of all the speaker pods. As this queries the API server, it waits at
 // least 5 minutes between queries, to avoid self-induced API server DoS (should
 // be trated with care).
-func (sl *SpeakerList) updateSpeakerIPs() {
+func (sl *SpeakerList) updateSpeakerIPs(stop <-chan struct{}) {
 	for {
 		// This either stops the loop or makes sure to wait at least 5
 		// minutes before making another call to sl.mlSpeakers(), which
 		// queries the API server.
 		select {
-		case <-sl.stopCh:
+		case <-stop:
+			sl.l.Log("op", "shutdown", "msg", "stopping updateSpeakerIPs")
+			sl.childrenWG.Done()
 			return
 		case <-time.After(5 * time.Minute):
 			// This blocks until the API server responds.
@@ -198,7 +209,7 @@ func (sl *SpeakerList) mlSpeakers() ([]string, error) {
 	return iplist, nil
 }
 
-func (sl *SpeakerList) joinMembers() {
+func (sl *SpeakerList) joinMembers(stop <-chan struct{}) {
 	// Every one minute, try to rejoin.
 	// This will join nodes that leave the cluster only for a few seconds every 1
 	// minute, while discovering new IPs (updated by sl.speakersIPsLoop()) might
@@ -207,7 +218,9 @@ func (sl *SpeakerList) joinMembers() {
 
 	for {
 		select {
-		case <-sl.stopCh:
+		case <-stop:
+			sl.l.Log("op", "shutdown", "msg", "stopping joinMembers")
+			sl.childrenWG.Done()
 			return
 		case <-ticker.C:
 			sl.mlJoin()
@@ -295,6 +308,11 @@ func (sl *SpeakerList) Stop() {
 		return
 	}
 
+	sl.l.Log("op", "shutdown", "msg", "stopping child goroutines")
+	close(sl.childrenStopCh)
+	sl.childrenWG.Wait()
+	sl.l.Log("op", "shutdown", "msg", "child goroutines stopped")
+
 	sl.l.Log("op", "shutdown", "msg", "leaving memberlist cluster")
 	err := sl.ml.Leave(time.Second)
 	sl.l.Log("op", "shutdown", "msg", "left memberlist cluster", "error", err)
@@ -306,13 +324,15 @@ func event2String(e memberlist.NodeEventType) string {
 	return []string{"NodeJoin", "NodeLeave", "NodeUpdate"}[e]
 }
 
-func (sl *SpeakerList) memberlistWatchEvents() {
+func (sl *SpeakerList) memberlistWatchEvents(stop <-chan struct{}) {
 	for {
 		select {
 		case e := <-sl.mlEventCh:
 			sl.l.Log("msg", "node event - forcing sync", "node addr", e.Node.Addr, "node name", e.Node.Name, "node event", event2String(e.Event))
 			sl.client.ForceSync()
-		case <-sl.stopCh:
+		case <-stop:
+			sl.l.Log("op", "shutdown", "msg", "stopping memberlistWatchEvents")
+			sl.childrenWG.Done()
 			return
 		}
 	}
```

We create a dedicated "children stop channel" for the SpeakerList's "background processes". We pass this channel to all child goroutines as an argument. When we receive a signal on the "main" stop channel, we close the children's channel and wait for the wait group.

We can probably tweak the logs a bit though.

Reply:
I've pushed the code here: https://github.com/kinvolk/metallb/commit/a6d84149ba6c13076face44af1acc75c20872b78